src/compiler/checker.ts

```diff
@@ -21286,6 +21286,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 tracing?.pop();
             }
 
+            // if (entry !== undefined) {
+            //     // If the previous entry and the result disagree, then something has gone wrong.
+            //     Debug.assert(!!(entry & RelationComparisonResult.Succeeded) === (result !== Ternary.False), "Cached relationship does not match recalculated result");
+            // }

```

Comment:
Probably this is a good thing to leave in, once the actual fix is in, that way we can observe this kind of bug faster.